atomic_reactor/plugins/post_flatpak_create_oci.py

```diff
@@ -0,0 +1,308 @@
+# add_app_prefix('org.gnome', 'eog, 'eog.desktop') => org.gnome.eog.desktop
+def add_app_prefix(app_id, root, full):
+    prefix = app_id
+    if prefix.endswith('.' + root):
+        prefix = prefix[0:-(1 + len(root))]
+    return prefix + '.' + full
+
+
+def find_desktop_files(builddir):
+    desktopdir = os.path.join(builddir, 'files/share/applications')
+    for (dirpath, dirnames, filenames) in os.walk(desktopdir):
+        for filename in filenames:
+            if not filename.endswith('.desktop'):
+                continue
+            yield os.path.join(dirpath, filename)
+
+
+def find_icons(builddir, name):
+    icondir = os.path.join(builddir, 'files/share/icons/hicolor')
+    for (dirpath, dirnames, filenames) in os.walk(icondir):
+        for filename in filenames:
+            if not filename.startswith(name + '.'):
+                continue
+            yield os.path.join(dirpath, filename)
+
+
+def update_desktop_files(app_id, builddir):
+    for full_path in find_desktop_files(builddir):
+        cp = ConfigParser.RawConfigParser()
+        cp.read([full_path])
+        try:
+            icon = cp.get('Desktop Entry', 'Icon')
+        except ConfigParser.NoOptionError:
+            icon = None
+
+        # Does it have an icon?
+        if icon and not icon.startswith(app_id):
+            found_icon = False
+
+            # Rename any matching icons
+            for icon_file in find_icons(builddir, icon):
+                shutil.copy(icon_file,
+                            os.path.join(os.path.dirname(icon_file),
+                                         add_app_prefix(app_id, icon, os.path.basename(icon_file))))
+                found_icon = True
+
+            # If we renamed the icon, change the desktop file
+            if found_icon:
+                subprocess.check_call(['desktop-file-edit',
+                                       '--set-icon',
+                                       add_app_prefix(app_id, icon, icon), full_path])
+
+        # Is the desktop file not prefixed with the app id, then prefix it
+        basename = os.path.basename(full_path)
+        if not basename.startswith(app_id):
+            shutil.move(full_path,
+                        os.path.join(os.path.dirname(full_path),
+                                     add_app_prefix(app_id,
+                                                    basename[:-len('.desktop')],
+                                                    basename)))
+
+
+class FlatpakCreateOciPlugin(PostBuildPlugin):
+    key = 'flatpak_create_oci'
+    is_allowed_to_fail = False
+
+    def __init__(self, tasker, workflow):
+        """
+        :param tasker: DockerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        """
+        super(FlatpakCreateOciPlugin, self).__init__(tasker, workflow)
+
+    def _get_target_path(self, export_path):
+        if self.source.runtime:
+            if export_path == "var/tmp/flatpak-build":
```

Comment:
This was a pretty confusing function.... I rewrote it for clarity, then it bothered me that it was doing a whole bunch of string concatenation and slicing for tens of thousands of files, so I rewrote it again. Took half an afternoon to get back to something that does the same thing. :-/